File: Sources/MalayalamPanchangamCalendar/Infrastructure/Calculation/MalayalamDateCalculator.swift
import Foundation

struct MalayalamDateCalculator: Sendable {
    // Sidereal mean solar motion: 360° / 365.25636 days (sidereal year)
    private let meanSiderealSolarMotionPerDay = 0.98560910

    func malayalamDate(for date: Date, siderealSunLongitude: Double, timeZone: TimeZone) throws -> (month: MalayalamMonth, day: Int, kollavarshamYear: Int) {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = timeZone
        let year = calendar.component(.year, from: date)

        let normalized = siderealSunLongitude.normalizedDegrees
        let month = monthForSiderealSunLongitude(normalized)
        let degreesIntoSign = normalized.truncatingRemainder(dividingBy: 30)
        // Malayalam months span one zodiac sign (≤31 days); cap at 31 to avoid invalid day 32
        let day = max(1, min(31, Int(floor(degreesIntoSign / meanSiderealSolarMotionPerDay)) + 1))
        let kollavarsham = kollavarshamYear(gregorianYear: year, malayalamMonth: month)
        return (month, day, kollavarsham)
    }

    private func monthForSiderealSunLongitude(_ longitude: Double) -> MalayalamMonth {
        let normalized = longitude.normalizedDegrees
        let zodiacSignIndex = Int(floor(normalized / 30.0))
        let monthRawValue = ((zodiacSignIndex + 8) % 12) + 1
        return MalayalamMonth(rawValue: monthRawValue) ?? .chingam
    }

    private func kollavarshamYear(gregorianYear: Int, malayalamMonth: MalayalamMonth) -> Int {
        switch malayalamMonth {
        case .chingam, .kanni, .thulam, .vrischikam, .dhanu:
            gregorianYear - 824
        case .makaram, .kumbham, .meenam, .medam, .edavam, .mithunam, .karkidakam:
            gregorianYear - 825
        }
    }
}

File: web/src/engine/MalayalamDateCalculator.ts
// ─────────────────────────────────────────────────────────────────────────────
// MalayalamDateCalculator
// Ported from MalayalamDateCalculator.swift (36 lines)
// ─────────────────────────────────────────────────────────────────────────────

import type { MalayalamMonthId } from '../models/MalayalamCalendar'

// Sidereal mean solar motion (degrees/day) — sidereal year basis
const SIDEREAL_SOLAR_MOTION = 0.98560910

export interface MalayalamDate {
  month: MalayalamMonthId
  day: number
  kollavarshamYear: number
}

/**
 * Compute the Malayalam date from the sidereal Sun longitude.
 * @param siderealSunLon  Sidereal Sun longitude in degrees [0, 360)
 * @param gregorianYear   The Gregorian year of this date (for KE year offset)
 * @param gregorianMonth  1-based Gregorian month
 */
export function malayalamDate(
  siderealSunLon: number,
  gregorianYear: number,
  gregorianMonth: number,
): MalayalamDate {
  // Malayalam month = sidereal zodiac sign (Leo = Chingam = 0)
  // Leo starts at 120° in the tropical zodiac, but since we use sidereal Sun,
  // we map: Chingam (Leo) = 120°, Kanni (Virgo) = 150°, ..., Karkidakam (Cancer) = 90°
  // The mapping is: month = floor(siderealSunLon / 30), then shift by Leo offset.
  // Tropical sign 0 = Aries; Malayalam month 0 = Chingam (Leo, sign 4).
  // sidereal Aries start = 0°. Leo = 120°.
  const signIndex = Math.floor(siderealSunLon / 30)  // 0=Aries .. 11=Pisces
  // Map: Aries(0) -> Medam(8), Taurus(1)->Edavam(9), ..., Leo(4)->Chingam(0)
  const SIGN_TO_MONTH: MalayalamMonthId[] = [8, 9, 10, 11, 0, 1, 2, 3, 4, 5, 6, 7]
  const month = SIGN_TO_MONTH[signIndex]

  // Day within the month = degrees into the current sign / mean daily motion
  const degreesIntoSign = siderealSunLon - signIndex * 30
  const day = Math.max(1, Math.min(31, Math.floor(degreesIntoSign / SIDEREAL_SOLAR_MOTION) + 1))

  // Kollavarsham (Kerala Era) year:
  // If Malayalam month is Chingam through Karkidakam (Leo–Cancer, i.e. Aug–Jul), KE = Gregorian - 825
  // Months after Chingam (≥1) that occur Aug-Dec use gregorianYear - 824
  // Simple rule used in the Swift app: months 0-7 (Chingam-Meenam) → gregorianYear - 824
  //   months 8-11 (Medam-Karkidakam) → gregorianYear - 825
  // Note: cross-check this with the original Swift logic.
  // Swift: Chingam-Dhanu (0-4) → year-824; Makaram-Karkidakam (5-11) → year-825
  const kollavarshamYear = month <= 4 ? gregorianYear - 824 : gregorianYear - 825

  return { month, day, kollavarshamYear }
}
